### ISSUE TITLE
Document the behavior and expectations of the various caches

### ISSUE THREAD

**Issue by Kukovec** [informalsystems/apalache#2362]
Currently, it's not very clear what the caches do and where they're written to/read from.

**Comment by thpani** [comment]
Also relevant for #2137

**Comment by konnov** [comment]
I am removing myself as an assignee in this issue, but happy to help if somebody takes a lead on this issue!

**Comment by thpani** [comment]
@konnov Honestly, I doubt that anyone but you can do this within reasonable time.

**Comment by konnov** [comment]
Yeah, but then it will not happen in a reasonable time frame. I just want to reflect reality here :)

**Comment by shonfeder** [comment]
As per our discussion last week, @Kukovec or @thpani are to update this issue with more specifics on what we need documented, and then we will get Igor to help fill in these identified gaps.

**Comment by thpani** [comment]
I think a good starting point, as mentioned by @Kukovec in the meeting, is to document

* what each cache is used for
* when each cache SHOULD be called
* when each cache MUST be called / what items can be expected to be in-cache
* when each cache is popped/cleared

**Comment by Kukovec** [comment]
Adding to the above, caches will also be an issue in follow-ups to #2547, since arena/SMT splitting would require us to reevaluate the caches, as the cached values are cells (arena), but the cache constraints are in SMT.
